DataStructures/HashTable/HashTable.cs

```diff
@@ -32,62 +38,81 @@ public bool Contains(object key)
         public void Add(object key, object value)
         {
             int index = Hash(key);
-            if (index <= _array.Length - 1 && _array[index] != null)
+            if (Contains(key))
             {
                 throw new DuplicateNameException();
             }
 
-            if (index > _array.Length - 1)
+            if (CheckForFull(index))
             {
-                Copy(_array, index);
+                _array = Resize(index + 1);
             }
 
-            _array[index] = value;
+            _array[index] = new ArrayValue();
+            _array[index].Value = value;
         }
 
-        public void Copy(object[] array, int index)
+        private bool CheckForFull(int index)
```

Comment:
IsResizeRequired instead of CheckForFull?

Reply:
or CheckSize

Reply:
Exactly)